acdc/tracr/utils.py

```diff
@@ -236,57 +238,115 @@ def get_perm(n, no_fp = True):
         perm = torch.randperm(n)
     return perm
 
-def get_tracr_data(tl_model, task: Literal["reverse", "proportion"]):
+def get_all_tracr_things(task: Literal["reverse", "proportion"], metric_name: str, num_examples: int, device):
+    _, tl_model = get_tracr_model_input_and_tl_model(task=task, device=device)
+    tl_model = tl_model.to(device)
+
     if task == "reverse":
         batch_size = 6
         seq_len = 4
-        data_tens = torch.zeros((batch_size, seq_len)).int()
+        data_tens = torch.zeros((batch_size, seq_len), device=device, dtype=torch.long)
+
+        if num_examples != batch_size:
+            raise ValueError("num_examples must be equal to batch_size for reverse task")
 
         vals=[0,1,2]
         import itertools
         for perm_idx, perm in enumerate(itertools.permutations(vals)):
             data_tens[perm_idx] = torch.tensor([3, perm[0], perm[1], perm[2]])
 
-        n = len(data_tens)
-        data_tens = data_tens.long()
-        patch_data_indices = get_perm(n)
+        patch_data_indices = get_perm(len(data_tens))
         warnings.warn("Test that this only considers the relevant part of the sequence...")
         patch_data_tens = data_tens[patch_data_indices]
         base_model_logprobs = F.log_softmax(tl_model(data_tens), dim=-1)
-        metric = partial(kl_divergence, base_model_logprobs=base_model_logprobs, mask_repeat_candidates=None)
+
+        if metric_name == "kl_div":
```

Comment:
Fairly certain this is updated in a later PR

Reply:
oh okay, I wasn't aware we wanted to do L2 for both tracr tasks. I can change it and re-run this task with l2.